app/javascript/react/components/molecules/Calendar/Calendar.tsx

```diff
@@ -1,27 +1,190 @@
+  const [dateReference, setDateReference] = useState<MovableCalendarData>({
+    zeroDate: "",
+    currentStartDate: "",
+    currentEndDate: "",
+    direction: undefined,
+    triggerDirectionUpdate: 0,
+  });
+
+  const updateMoveValue = (direction: MovesKeys) => {
+    setDateReference((prevState) => {
+      return {
+        ...prevState,
+        direction: direction,
+        triggerDirectionUpdate: prevState.triggerDirectionUpdate + 1,
+      };
+    });
+  };
+
+  const handleBackwardMovement = (dateMoment: moment.Moment) => {
+    const newEndMoment = dateMoment.subtract(1, "months").endOf("month");
+    setIsButtonDisabled(false);
+    return newEndMoment;
+  };
+
+  const handleForwardMovement = (dateMoment: moment.Moment) => {
+    let newEndMoment = dateMoment.add(1, "months").endOf("month");
+    const zeroDateMoment = moment(dateReference.zeroDate, "YYYY-MM-DD");
+    if (newEndMoment.isAfter(zeroDateMoment)) {
+      newEndMoment = zeroDateMoment;
+      setIsButtonDisabled(true);
+    }
+    return newEndMoment;
+  };
+
+  useEffect(() => {
+    // Disable forward button at first, becouse
+    // we present the newest months just after loading this page
+    setIsButtonDisabled(true);
+
+    const lastElementIdx = movingCalendarData.data.length - 1;
+    const endDate = movingCalendarData.data[lastElementIdx].date;
+
+    const startMoment = moment(endDate, "YYYY-MM-DD");
+    const newStartDate = startMoment
+      .date(1)
+      .subtract(SEEN_MONTHS_NUMBER - 1, "months")
+      .format("YYYY-MM-DD");
+
+    setDateReference((prevState) => ({
+      ...prevState,
+      zeroDate: endDate,
+      currentStartDate: newStartDate,
+      currentEndDate: endDate,
+    }));
+  }, []);
+
+  useEffect(() => {
+    if (!dateReference.currentEndDate) return;
+
+    const dateMoment = moment(dateReference.currentEndDate, "YYYY-MM-DD");
+    let newEndMoment: Moment;
+
+    switch (dateReference.direction) {
+      case -1:
+        newEndMoment = handleBackwardMovement(dateMoment);
+        break;
+      case 1:
+        newEndMoment = handleForwardMovement(dateMoment);
+        break;
+      default:
+        console.error("Invalid direction value:", dateReference.direction);
+        return;
+    }
+
+    const newEndDate = newEndMoment.format("YYYY-MM-DD");
+
+    const newStartDate = newEndMoment
+      .date(1)
+      .subtract(SEEN_MONTHS_NUMBER - 1, "months")
+      .format("YYYY-MM-DD");
+
+    setDateReference((prevState) => ({
+      ...prevState,
+      currentStartDate: newStartDate,
+      currentEndDate: newEndDate,
+    }));
+
+    dispatch(
+      fetchNewMovingStream({
+        id: streamId,
+        startDate: newStartDate,
+        endDate: newEndDate,
+      })
+    );
+  }, [dateReference.triggerDirectionUpdate]);
+
+  const handleLeftClick = () => {
+    updateMoveValue(MovesKeys.MOVE_BACKWARD);
+  };
+
+  const handleRightClick = () => {
+    updateMoveValue(MovesKeys.MOVE_FORWARD);
+  };
```

Comment:
Could we move the logic for handling date changes and rendering buttons to another file?  WDT?